_docs/tutorial/05-adding-a-new-version-of-a-microservice.md

```diff
@@ -0,0 +1,71 @@
+     ```bash
+     kubectl exec -it $(kubectl get pod -l app=sleep -o jsonpath='{.items[0].metadata.name}') bash
+     for i in {1..10}; do curl -o /dev/null -s -w "%{http_code}\n" <the value of DETAILS_V2_POD_IP>:9080/reviews/7; done
+     ```
+4. Now we are rather confident that our new version of _reviews_ will work and we will release it. We will release a single replica of it into production - the real production traffic will arrive to our new version. In the current setting, 75% of the traffic will arrive to the old version (three pods of the old version) and 25% will arrive to the new version (one pod).
+
+  To release _reviews v2_ we will undeploy our new version and redeploy it with the _app_ label, so it will become addressable by the _reviews_ service.
+
+  ```bash
+  kubectl delete -f bookinfo-reviews-v2-without-app-label.yaml
+  kubectl apply -f bookinfo-reviews-v2-with-app-label.yaml
+  ```
+
+  We can check the labels of our pod:
+  ```
+  kubectl get pods --show-labels -l version=v2
+  ```
+5. Now we access the application web page and see that the black stars appear for ratings! We will hit the page several times and see that the sometimes the page is returned with stars (approximately 25% of the time) and sometimes without stars (approximately 75% of the time)
+
+6. If we encounter any problems we can quickly undeploy the new version, so only the old version will be used.
+
+7. Next we will increase the replicas of our new version. We can do it gradually, carefully checking that our number of errors does not increase:
+  ```bash
+  kubectl scale deployment reviews-v2 --replicas=3
+  ```
+8. Now we decommission the old version
+  ```bash
+  kubectl delete deployment reviews-v1
+  ```
+
+9. Accessing the web page of the application will return reviews with black stars only.
+
+We performed the update of _reviews_ pretty well. First, we deployed the new version without directing to it any production traffic. We tested it in the production environment, on test traffic. We checked that the new version does not produce any errors. We released the new version, gradually increasing the production traffic to it. Finally, we decommissioned the old version.
+
+It all went good, however we want to improve our release strategy. First, we want to allow our testers to test the new version end-to-end in production. For that we need an ability to drive the traffic to our new version by request parameters, for example by the user name stored in a cookie. Kubernetes cannot provide this routing functionality. Also, we would like to perform _shadowing_ of the production traffic to our new version and checking if our new version provides incorrect results or produces any errors. Kubernetes cannot do it. Also, we would like to be more fine grained with our rollout. We would like to release our new version to 10% of the users and then increase it by 10%. Kubernetes cannot be this fine-graned.
```

Comment:
went good -> went well
Kubernetes cannot do it -> Kubernetes can't do that either.
fine grained -> fine-grained

Reply:
Maybe mention all three things, and then just say once "Kubernetes is unable to directly help with any of these tasks".